extkingpin/pathorcontent_reloader.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) The EfficientGo Authors.
+// Licensed under the Apache License 2.0.
+
+// Taken from Thanos project.
+//
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package extkingpin
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+)
+
+// logger is an interface compatible with go-kit/logger.
+type logger interface {
+	Log(keyvals ...interface{}) error
+}
+
+// pathOrContent is an interface compatible with PathOrContent.
+type pathOrContent interface {
+	Content() ([]byte, error)
+	Path() string
+}
+
+// PathContentReloader starts a file watcher that monitors the file indicated by pathOrContent.Path() and runs
```

Comment:
Let's make a note this can run forever if not cancelled?

Also I wonder if it wouldn't be cleaner to make it sync. Otherwise it looks like a function that just leaks things when you are not careful. So either make it sync so others would run it in go routine or return closer?